### DISCUSSION TITLE
Featherpad doesn't get raised/focused when opening a file in a new tab.

### DISCUSSION THREAD

**Discussion by HubKing** [tsujan/FeatherPad#797] [Q&A] [answered]
Plasma 6.0.4. The default text editor is Feather. In Dolphin file manager, if I double click a text file, Feather starts with that file and gets raised/focused. But If I double-click another text file in Dolphin, Feather doesn't get raised/focused. Instead, it only opens a new tab with that file. This is problematic for me, if Feather is hidden beneath other windows or minimised. Is this an intended behaviour, or a bug of Plasma?

If it's related, my window behaviour is like below.
![image](https://github.com/tsujan/FeatherPad/assets/10900489/c946f7cf-0bd8-499f-bd17-387c26c22a3f)

**Comment by tsujan** [comment, marked as answer]
If you use X11, this is about what is usually called WM's "focus stealing prevention". Since it's set to "Low" in your screenshot, you're probably under Wayland, not X11.

But if you use Wayland, no app can activate, raise or position its window under Wayland.

In short, it's about the WM, not the app.

**Reply by HubKing** [reply]
So, under X11, in this case, Feather with new tab gets raised/focused?

**Reply by tsujan** [reply]
Yes, under X11, FeatherPad uses all the available Qt tricks to circumvent the focus stealing prevention of the WM, unless the latter is very high.

I work under LXQt+LabWC (Wayland) most of the time and because of this, I enable `Preferences → Window → Always open in separate windows`. When I log into LXQt+kwin_x11, not only the window is activated and focused, but also the app can know about virtual desktops and open a new window only when there's none on the current desktop. Wayland doesn't let apps do these things, but for other reasons, I prefer it to X11.

**Reply by HubKing** [reply]
Well, this goes to my list of the lacking functionalities of Wayland that makes me think that the designers of Wayland are out of touch with the reality of desktop. X11 is on life-support, and everyone is moving to Wayland, but there are so many features lacking in Wayland that all other desktop OS's have, like always-on-top,or specifying window positions, etc. Maybe this is also for "security", but I feel like it's as if making no window on a house, because bugs can enter through windows.

**Reply by tsujan** [reply]
> that makes me think that the designers of Wayland are out of touch with the reality of desktop

I agree, and I knew it more than a decade ago ;) If you asked me about Wayland just a year ago, I might say it was ridiculous (actually, I've said similar things in the last decade, by explaining the reasons). There are always fanboys that think something is good just because it's new, without knowing what it is.

That being said, there have been diligent, rational developers who have tried to make Wayland really usable. I admire the work that is (being) done in `wlroots` and Wayland compositors that are based on it, like LabWC and Wayfire (especially the first one). KDE devs have also had a big role in making Qt work as fine as possible under Wayland, and their work isn't limited to KDE at all.

So, now, I think Wayland *is* good enough because (1) it's lighter and more efficient than X11 in drawing windows, and (2) good developers are making it more usable.

**Reply by HubKing** [reply]
Seems like FireFox (Wayland) can at least highlight the task bar entry when I clicked an HTTP link and FireFox opened it as a new tab in an existing background window, even though it doesn't raise/focus the window like under Windows. 

![image](https://github.com/tsujan/FeatherPad/assets/10900489/6f3f7372-5dac-4bbd-8c24-fd391da38b83)

**Reply by HubKing** [reply]
Seems like this is easy to implement. All I had to do was just calling [`QApplication::alert()`](https://doc.qt.io/qt-6/qapplication.html#alert). Feather could call it when it opens a new tab by the system's request and its window is in the background.

![image](https://github.com/tsujan/FeatherPad/assets/10900489/b81c0ed7-c4ab-4d9b-9138-626a9af95bc6)

**Reply by tsujan** [reply]
Yes, but that's what kwin_wayland and the Plasma task-bar support. It isn't a Wayland property, and not all Wayland compositors support it.

However, I agree that it may be good to add under Wayland; will experiment with it under various Wayland compositors and with different task-bars.

**Reply by tsujan** [reply]
As I'd suspected, it only worked under Plasma-Wayland; nothing happened under LabWC or Wayfire. There may be Qt-specific codes for it in Plasma, which is understandable.

I might discuss it with LabWC devs though. They're very active and care for Qt. We should see whether it has a connection to Wayland through QtWayland.

**Reply by stefonarch** [reply]
I've a personal workaround which works for the filemanager but not if using `fpad file.txt`:
```
Exec=bash -c "featherpad %U && wlrctl window focus featherpad"
``` 
in  `~/.local/share/applications/featherpad.desktop`

**Reply by tsujan** [reply]
Yes, but here the question is, "Can Qt's activation *demanding* have effect with `wlroots`?" We know that activation itself doesn't have any effect.

**Reply by tsujan** [reply]
Anyway, I added a commit: https://github.com/tsujan/FeatherPad/commit/11db762be65a89ba08981c4e598ab91ba9611525. At least it works under Plasma-Wayland.

**Comment by tsujan** [comment]
There's only one step from being sensitive to attention demands and really activating windows according to a "focus-stealing-prevention" scheme. I wonder if KDE devs have such a plan. With it, a great limitation of Wayland (which is imposed in the name of "security") will be bypassed.
